src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp

```diff
@@ -5311,6 +5313,365 @@ class StubGenerator: public StubCodeGenerator {
+      if (is_signed_subword_type(eltype)) {
+        __ sshll2(vhalf0, Assembler::T4S, vdata0, Assembler::T8H, 0);
+      } else {
+        __ ushll2(vhalf0, Assembler::T4S, vdata0, Assembler::T8H, 0);
+      }
+      __ addv(vmul0, Assembler::T4S, vmul0, vhalf0);
+    }
+
+    __ br(Assembler::HI, SMALL_LOOP);
+    guarantee(__ offset() - start == small_loop_size, "Incorrect small_loop_size");
+
+    // SMALL LOOP'S EPILOQUE
+
+    __ lsr(rscratch2, cnt, exact_log2(evf));
+    __ cbnz(rscratch2, LARGE_LOOP_PREHEADER);
+
+    __ mulv(vmul0, Assembler::T4S, vmul0, vpow);
+    __ addv(vmul0, Assembler::T4S, vmul0);
+    __ umov(result, vmul0, Assembler::S, 0);
+
+    // TAIL
+
+    const int tail_size = (8 + (vf - 1) * 2) * 4; // 14 or 22 insts
+    if (tail_size % 32 > 32 - __ offset() % 32) {
+      __ align(32);
+    }
+
+    __ bind(TAIL);
+    start = __ offset();
+    assert(is_power_of_2(vf), "can't use this value to calculate the jump target PC");
+    __ andr(rscratch2, cnt, vf - 1);
+    __ bind(TAIL_SHORTCUT);
+    __ adr(rscratch1, RELATIVE);
+    __ sub(rscratch1, rscratch1, rscratch2, ext::uxtw, 3);
+    __ movw(rscratch2, 0x1f);
+    __ br(rscratch1);
+
+    for (size_t i = 0; i < vf - 1; ++i) {
+      large_arrays_hashcode_elload(rscratch1, Address(__ post(ary, type2aelembytes(eltype))),
+                                   eltype);
+      __ maddw(result, result, rscratch2, rscratch1);
+    }
+    __ bind(RELATIVE);
+
+    __ leave();
+    __ ret(lr);
+    guarantee(__ offset() - start == tail_size, "unexptected size of the tail code block");
+
+    // LARGE LOOP
+
+    __ align(32);
+    __ bind(LARGE_LOOP_PREHEADER);
+
+    __ lsr(rscratch2, cnt, exact_log2(evf));
+
+    if (multiply_by_halves) {
+      // 31^4 - multiplier between lower and upper parts of a register
+      __ movw(rscratch1, intpow(31U, vf / 2));
+      __ mov(vpowm, Assembler::S, 1, rscratch1);
+      // 31^28 - remainder of the iteraion multiplier, 28 = 32 - 4
+      __ movw(rscratch1, intpow(31U, evf - vf / 2));
+      __ mov(vpowm, Assembler::S, 0, rscratch1);
+    } else {
+      // 31^16
+      __ movw(rscratch1, intpow(31U, evf));
+      __ mov(vpowm, Assembler::S, 0, rscratch1);
+    }
+
+    __ mov(vmul3, Assembler::T16B, 0);
+    __ mov(vmul2, Assembler::T16B, 0);
+    __ mov(vmul1, Assembler::T16B, 0);
+
+    const int large_loop_size = load_arrangement == Assembler::T4S   ? 44  // 11 insts
+                                : load_arrangement == Assembler::T4H ? 60  // 15 insts
+                                : load_arrangement == Assembler::T8H ? 108 // 27 insts
+                                : load_arrangement == Assembler::T8B ? 124 // 31 insts
+                                                                     : -1; // invalid
+    guarantee(large_loop_size != -1, "invalid small_loop_size");
+
+    if (large_loop_size % 32 > 32 - __ offset() % 32) {
+      __ align(32);
+    }
+
+    start = __ offset();
```

Comment:
What does this logic from 5552 onwards do? It at least deserves a comment.
Can we not simply delete it?

Reply:
It's here to make sure the loop takes the smallest number of aligned 32-byte instruction memory regions possible.

Reply:
Does that make a significant measurable difference?
Why not simply 32-align the region? Then we can get rid of this `large_loop_size` calculation.

Reply:
> Does that make a significant measurable difference?

I'll revert on this with performance numbers later.

> Why not simply 32-align the region? Then we can get rid of this large_loop_size calculation.

We aim to align the code only if it reduces the number of aligned 32-byte instruction memory regions the loop compromises.

Reply:
If doing this really does help performance for this function, then we can find a good way to do it which doesn't require any lengths to be hard coded, and we can fully document the solution so others can use it.

Reply:
Doing this gives no performance benefits. Removed by https://github.com/openjdk/jdk/pull/18487/commits/9f4ae8554106d3e1290ee127b848d7f48952e5fb .